subxt/src/client/lightclient/platform/native.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
```

Comment:
I've already asked this question to the `cumulus` team, and I'm repeating it here: why rewrite this implementation of `Platform`?

Smoldot [provides an implementation](https://github.com/smol-dot/smoldot/blob/main/light-base/src/platform/default.rs), and https://github.com/paritytech/subxt/pull/965 uses it. Why decide that copy-pasting its code here (without caring for the licensing, but let's forget that) when there's already a fully working and maintained version provided by smoldot itself?

To me you (and Cumulus) are really shooting yourself in the foot by doing this, and I just don't understand it.

We all know that this code will rot over time. You have the choice between the simple option that will work forever, and the complicated option that will likely require effort to maintain, and you go for the complicated one.

Reply:
Hey, I've implemented the low level `Platform` here to expose a pure-rust based implementation for wasm. While at it, I've also ported the platform for native (using tokio sockets) to allow us to reuse the same code for both platforms for better coverage. The differences are mainly in the `connect` function which uses a custom websocket with minimal dependencies that also implements `AsyncRead` / `AsyncWrite`.


Yep, I also felt the frustration with the `wasm-bindgen`, fortunately, we are using it for the custom socket. The socket is an oversimplification of [gloo-net](https://github.com/rustwasm/gloo/blob/aa0dbcdf3e27fdf359efe1259e5021e6e20fd3fb/crates/net/src/websocket/futures.rs#L48) (this is what we use by default with jsonrpsee for wasm environments).
I'm not quite sure if we could have a pure-rust based implementation without some types of bindings for the websocket 🤔 I would be up for transitioning to something different

> I also want to point out that the code in this PR doesn't support WebRTC.

Yep, we'd need to maintain the platform in the meanwhile if we pick this road in the end and look for alternatives of webrtc

Reply:
@lexnv what are the major blockers to using @tomaka's `Platform` implementation? I agree that it would be nicer to rely on an implementation that was supported and kept uptodate for us, so would be good to understand exactly the issues with doing so :)

Reply:
Pierre's platform works perfectly for native environments and there's also an alternative for WASM.
To make it fully compatible with our WASM needs the functions from [lightclient/src/platform/wasm.rs](https://github.com/paritytech/subxt/pull/1026/files#diff-1aa5737258d86a40987b2ab081bf272a5960e7051d042ce41e954ab933df57a6) are added to this PR.
For WASM we need: `spawn`, `now_from_unix_epoch`, `now`, `sleep`, and most importantly `connect` together with a few types, while for native we (including Cumulus) need just the `connect` to use a different socket.

Maybe we could leverage 90% of the smoldot Platform by having an extension trait.
The parts of the `Platform` that handles buffers are common between environments (`lightclient/src/platform/default.rs`), and the trickiest to implement right. Then we would need to implement just the connect for native and avoid duplicating the code here and in Cumulus.

```rust
// Extension trait to reuse the common parts of the Platform that handles buffers internally
trait PlatformRefBuffered {
 type Instant;
 type Delay;
 
 fn spawn() ..
 fn now() ..
 }

impl<T: PlatformRefBuffered> Platform for T {
...
}
```

### Smoldot Wasm Node

There's a crate in smoldot that implements this.
The crate requires the javascript realm to expose these [bindings](https://github.com/smol-dot/smoldot/blob/c300c082ec34a749108e71f12912e1355e168c67/wasm-node/rust/src/bindings.rs#L63-L64).

```rust
#[link(wasm_import_module = "smoldot")]
extern "C" {
  pub fn buffer_copy(buffer_index: u32, target_pointer: u32);
  ...
```

These are expected to be placed under the linking module "smoldot".
Before manually implementing the `Platform` here I also tried to create these bindings in rust, then reuse the wasm platform from this crate.

However, I was unable to expose any external functions under the `smoldot` module.
The WAT binary would expect the bindings as:

```bash
(import "smoldot" "buffer_copy" (func $_ZN18smoldot_light_wasm8bindings11buffer_copy17hf21d3e4eee7463c1E (type 6)))
(import "smoldot" "advance_execution_ready" (func $_ZN18smoldot_light_wasm8bindings23advance_execution_ready17h100f38f424c0220dE (type 0)))
(import "smoldot" "connection_new" (func $_ZN18smoldot_light_wasm8bindings14connection_new17hab2362b6b1ccdf39E (type 28)))
```
While regardless of where I tried to implement the functions they would be exported from rust as

```bash
(export "buffer_copy" (func $buffer_copy))
(export "buffer_size" (func $buffer_size))
(export "json_rpc_responses_non_empty" (func $json_rpc_responses_non_empty))
```

We could overcome this eventually by modifying smoldot and removing `#[link(wasm_import_module = "smoldot")]`.
However, all the binding functions are designed to work within the JavaScript world.
For example, the `buffer_copy` function takes as parameter `buffer_index: u32` which would mean we'd also need to implement in Rust a way to map a buffer's memory region to an index.
There's an example in [timers.rs](https://github.com/smol-dot/smoldot/blob/c300c082ec34a749108e71f12912e1355e168c67/wasm-node/rust/src/timers.rs) from the wasm-node, which uses some static variables to handle the mappings.

Just for reference, [here](https://github.com/smol-dot/smoldot/tree/main/wasm-node/javascript) is the JavaScript counterpart of the wasm platform.
An alternative to this would be writing the JavaScript code then binding to it in rust. Or trying to bind directly to the JavaScript code from wasm-platform.

That felt a bit more involved than having our slim socket bindings and using that common parts of the platform and I would be up for adjusting/modifying this to make things better for us 🙏

Reply:
> The parts of the Platform that handles buffers are common between environments (lightclient/src/platform/default.rs), and the trickiest to implement right

What I'm suggesting is to simply de-duplicate the code.

Sure the code is the same as in smoldot, but the choice here is between:

- Copy-pasting part of smoldot's `Platform` implementation into subxt and use it for both native and wasm sockets.
- Copy-pasting part of smoldot's `Platform` implementation into subxt and use only for wasm sockets.

I don't really see how solution 1 (which you're going for) is in any way better than solution 2.

> Before manually implementing the Platform here I also tried to create these bindings in rust,

That's not possible and I don't think it's a good approach nonetheless.

My personal intuition about the way forward is that I should try improve the `Platform` trait to be easier to implement, and/or maybe allow easily reusing smoldot's JavaScript.
But I can't really guarantee in any way that I'll succeed here.

Reply:
That totally makes sense! 🙏 

Indeed removing the native Platform from here and keeping only the WASM part will simplify the code for us.
The main advantage here would be that most users will be using a fully tested Platform provided by smoldot that we can easily update.
I was considering the native Platform from this PR to lay the foundation for testing and give us a bit more confidence, but I think we could overcome that if we get manage to get a bit more WASM testing.
And since we'll release this under an experimental/unstable feature flag in the beginning, I think we are fine for this PR.

> My personal intuition about the way forward is that I should try improve the Platform trait to be easier to implement, and/or maybe allow easily reusing smoldot's JavaScript.

Thanks! That would help us indeed! Just a crazy idea, maybe initially having a small platform extension for native only would also simplify the work from Cumulus side. I think Cumulus just needs a way to change the socket from `fn connect`

Reply:
> if we get manage to get a bit more WASM testing

IMO you need WASM-specific testing anyway, because browser APIs have many weird corners. I don't think it's a good idea to assume that things "just work" if you just plug APIs together.

> Thanks! That would help us indeed! Just a crazy idea, maybe initially having a small platform extension for native only would also simplify the work from Cumulus side. I think Cumulus just needs a way to change the socket from fn connect

I'm planning to rework the `connect` function quite a lot, for example to pass an already-parsed multiaddr, and to let smoldot do the WebSocket implementation on top of TCP (in other words the native implementation would just need to support TCP). But again these are just ideas and there are many more things on my to-do list that have priority.